### ISSUE TITLE
help tracking thread explosions?

### ISSUE THREAD

**Issue by danpf** [dask/distributed#2398]
I'm having a super weird problem with dask on my SLURMcluster using dask-jobqueue but i'm confident that's not the problem.

Actual problem:
Slurm is unable to submit jobs to some nodes because my jobs are generating a highly excessive number of processes(threads)

brief workflow
1. submit script to slurm
2. script from(1) generates a SLURMcluster (aka a dask client)
3. (2) submits to SLURM and makes the actual workers

Description of problem:
I will attatch 2 logfiles from `pstree -l -c -u danpf -a`
logfile `normal_op.txt` is from normal operation
Could someone help me figure out what all of these extra python processes are?
why does each forkserver need 6 threads, and why does each worker need 13 threads?
This is a node processer load load of approx 13 or so per worker thread. this is pretty high. Ideally it would be around 6-7 (or so my admin tells me, they would prefer it to be as low as possible)

However, after a worker fails, which seems to occur for various reasons (including but not limited to):
1. the main scheduler encountering an exception and dying
2. a SIGKILL command
3. a worker dying for some reason ( not too sure I get a lot of weird dask errors sometimes)

If i'm ssh'd into a node and i track the number of processes  / threads when this happens, I see an explosion of the number of threads from ~300-500 to ~7-8000. The result from `pstree` command from above after i've killed the master(main Client) thread is `cancel_op.txt`

At this point, I can't even ssh into a node because there's too many processes, and any time SLURM tries to submit something to that node, the command it submits dies almost instantly. (I'm unaware if this is limited to my submissions, or if i'm actually causing other people's submissions to die as well)

I guess the questions I  would like answered /help answering are:
1. what are all of these threads in the first place (normal per-worker load appears to be ~13 threads/processes) (is there a way to slim this down?)
2. what could be happening to cause these 'forkserver`   processes  to generate so many threads when they lose contact with the master process?
3. Would a different multiprocessing-method help with this? it wouldn't appear so from what i've read, but figured I would ask (it looks like only spawn, fork, and forkserver are options)

[normal_op.txt](https://github.com/dask/distributed/files/2646408/normal_op.txt)
[cancel_op.txt](https://github.com/dask/distributed/files/2646409/cancel_op.txt)

Thanks for any help!!!
Dan

**Comment by jhamman** [comment]
First thing I would check is if your workers are mixing with OpenMP parallelization. I would try setting OpenMP threads to 1. 

```yaml
jobqueue:
  slurm:
    env_extra:
      - 'export OMP_NUM_THREADS=1'
```

Generally, it would be good to know how you are configuring your workers. If this ends up being a jobqueue issue, we may consider moving this issue to https://github.com/dask/dask-jobqueue/issues. 

I might also suggest looking at http://jobqueue.dask.org/en/latest/debug.html for some debugging pointers.

**Comment by danpf** [comment]
if you limit yourself to a single node (as i'm showing above with 'node43')
this the the plot you get when you track threads over time: (starting before i even start the main Client)

![image](https://user-images.githubusercontent.com/16297104/49480351-3f7f3580-f7db-11e8-97cb-f4f93dd76fa6.png)

**Comment by danpf** [comment]
Did a little digging and the problem code somewhere here:

However, this is the deepest I can dig, I understand asyncio but I don't understand tornado, i'm guessing it has to do with the way errors are handled with the futures and the `gen.with_timeout function`

Maybe i can pass the torch on here? or someone could help with this. I think if you know what you're doing this is a pretty narrow scope.

```
    while True:
        try:
            print("in connect FINALLY")
            future = connector.connect(loc, deserialize=deserialize,
                                       **(connection_args or {}))
            comm = yield gen.with_timeout(timedelta(seconds=deadline - time()),
                                          future,
                                          quiet_exceptions=EnvironmentError)
        except EnvironmentError as e:
            print("found env error")
            error = str(e)
            if time() < deadline:
                yield gen.sleep(0.01)
                logger.debug("sleeping on connect")
            else:
                _raise(error)
        except gen.TimeoutError:
            print("found timeout")
            _raise(error)
        else:
            print("BROKE")
            break
        print("looping", counter)
        counter += 1
```

logs:
```
found env error                                
looping 0                                      
in connect FINALLY                             
found env error                                
looping 1                                      
in connect FINALLY                             
found env error                                
looping 2                                      
...
found env error                                
looping 419                                    
in connect FINALLY                             
found env error                                
looping 420                                    
in connect FINALLY                             
found env error                                
looping 421                                    
in connect FINALLY                             
```
while this is printing the process/thread count skyrockets.

**Comment by mrocklin** [comment]
I'm curious, can you provide any more information about the threads?  Do they happen to have names for example?  (take a look at `threading.enumerate()`).  I think that identifying where these threads came from is probably the right first step in diagnosing this issue.

**Comment by danpf** [comment]
if in one terminal i type 
`while true; do ps -u danpf -T | wc -l; done`
and in the other terminal i type
```
python -m distributed.cli.dask_worker tcp://localhost:28400 --nthreads 1 --memory-limit 4.00GB --name master_standard--d-- --death-timeout 60 --local-directory /local
```
i can watch the thread count rise by ~200 in the first terminal

thats the first few lines if I print out 
```
frame = sys._current_frames().get(thread.ident, None)
         if frame:
             return frame.f_code.co_filename, frame.f_code.co_name, frame.f_code.co_firstlineno
```

```
('/home/danpf/git/distributed/distributed/comm/core.py', 'get_thread_position', 182)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/multiprocessing/connection.py', '_recv', 374)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/linecache.py', 'updatecache', 82)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/selectors.py', 'select', 402)
('/home/danpf/git/distributed/distributed/comm/core.py', 'get_thread_position', 182)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/multiprocessing/connection.py', '_recv', 374)
('/home/danpf/git/distributed/distributed/profile.py', '_watch', 230)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/selectors.py', 'select', 402)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/concurrent/futures/thread.py', '_worker', 66)
('/home/danpf/git/distributed/distributed/comm/core.py', 'get_thread_position', 182)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/multiprocessing/connection.py', '_recv', 374)
('/home/danpf/git/distributed/distributed/profile.py', '_watch', 230)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/selectors.py', 'select', 402)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/concurrent/futures/thread.py', '_worker', 66)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/concurrent/futures/thread.py', '_worker', 66)
('/home/danpf/git/distributed/distributed/comm/core.py', 'get_thread_position', 182)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/multiprocessing/connection.py', '_recv', 374)
('/home/danpf/git/distributed/distributed/profile.py', '_watch', 230)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/selectors.py', 'select', 402)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/concurrent/futures/thread.py', '_worker', 66)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/concurrent/futures/thread.py', '_worker', 66)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/concurrent/futures/thread.py', '_worker', 66)
('/home/danpf/git/distributed/distributed/comm/core.py', 'get_thread_position', 182)
('/home/danpf/.local/share/pyenv/versions/3.7.0/lib/python3.7/multiprocessing/connection.py', '_recv', 374)
('/home/danpf/git/distributed/distributed/profile.py', '_watch', 230)
```
(get_thread_position is my function)
I'm not exactly sure if _watch really makes sense to be the culprit here though...

**Comment by danpf** [comment]
@mrocklin is there any documentation on what uses threads just in general?
I usually see ~14-20 threads per a process, and like memory or cpu usage, the total number of threads is a resource that is important to track, and 14-20 already is quite high, and on systems with 64 cores, that puts dask at using over 1000 threads which is well over most system's limits.

**Comment by mrocklin** [comment]
https://stackoverflow.com/questions/52618436/what-threads-do-dask-workers-have-active?r=SearchResults

**Comment by danpf** [comment]
@mrocklin one weird thing i'm noticing is that some workers end up with ~100 threads or so... I'm confident the threads aren't from OMP or BLAS, i'm wondering if incoming tasks are stored in separate threads, or if it will take more digging to figure that out...

also ps. in your question it should be `THREADS` not `THEADS`

**Comment by mrocklin** [comment]
> also ps. in your question it should be THREADS not THEADS

Thanks.  Fixed.

**Comment by mrocklin** [comment]
>  i'm wondering if incoming tasks are stored in separate threads

Dask workers only intentionally open the threads listed in that SO question.  As you've showed in https://github.com/dask/distributed/pull/2403 tornado/asyncio may also open up other threads.  To test for things like this I would print out the names of the threads using `threading.enumerate` and hope that that leads to some clue.  If threading.enumerate doesn't list much, then it's probably that they're opened by some other non-Python library.

**Comment by danpf** [comment]
i think mostly fixed by  #2403
